### ISSUE TITLE
BUG: Messages shown by emotes are only shown client-side

### ISSUE THREAD

**Issue by jadebenn** [DarkflameUniverse/DarkflameServer#1342]
### Make sure you've done the following:

- [X] I have checked that this issue has not already been reported.

- [X] I have validated that this issue is not a syntax error of either MySQL or SQLite.

- [X] I have pulled the latest version of the main branch of DarkflameServer and have confirmed that the issue exists there.

### DarkflameServer Version

e239166

### Issue Description

When playing an emote, the animation is shown to other players, but emotes that play messages do not show any messages to other players.

### Reproduction Steps

1. Get two players
2. Have one play an emote that should show a message
3. Observe the other player cannot see any message

### Expected Behavior

Other players should see a message above other players' heads when they play an emote that displays one.

### Environment

WSL

**Comment by aronwk-aaron** [comment]
I think this has something to do with chat messages and the local_emote channel.
I'm not sure if that even sends a chat message to the server for it and we just ignore it cause of the channel it comes in on, or if we should just know to send the chat from the emote GM that the server receives

**Comment by ElectScholar** [comment]
Has this issue been processed yet?

**Comment by EmosewaMC** [comment]
Not yet, someone would want to dig into packet capture to figure out how this is done and it hasnt been done yet.  If you'ld like to work on this i can do the necessary research and then leave the implementation to you @ElectScholar

**Comment by ElectScholar** [comment]
@EmosewaMC yes, that sounds great! I will also do some research later today about packet capture

**Comment by ElectScholar** [comment]
Sorry for the late replies, I can get started on this tonight!

**Comment by ElectScholar** [comment]
I tested the problem just to check and confirmed that emotes do play animations but do not show the corresponding messages associated with them.

I followed implementation instructions yet believe I am making an error becuase I don't understand the logic behind these bitstream.write() instances and how the server uses them to interact with the clients. I tried copying other message structures from the header file and for my Messagetype I used 'ADD_MESSAGE'. I wish to know more about this so if anyone knows more about how this logic works feel free to hit me up on discord server (revvverb) to chat for a bit and help me understand. 

My next session I will debug this further but as of right now the new code does not resolve the bug.

This is what I have so far:
`GameMessages.h`
```cpp
	struct EmotePlayed : public GameMsg {
		EmotePlayed() : GameMsg(MessageType::Game::ADD_MESSAGE) {}

		void Serialize(RakNet::BitStream& stream) const override;

		int32_t emoteID {-1};
		LWOOBJID targetID{ LWOOBJID_EMPTY };
	};
```
`GameMessages.cpp`
```cpp
	void EmotePlayed::Serialize(RakNet::BitStream& stream) const {
		stream.Write(emoteID != -1);
		if (emoteID != -1) stream.Write(emoteID);
		stream.Write(targetID != LWOOBJID_EMPTY);
		if (targetID != LWOOBJID_EMPTY) stream.Write(targetID);
		//since message has unassigned address we use SEND_PACKET_BROADCAST?
		SEND_PACKET_BROADCAST;
	}
```

Implementation:
`RenderComponent.cpp`
```cpp
	4980 | //RenderComponent::PlayAnimation(entity, sAnimationName);
	4981 | GameMessages::EmotePlayed(emoteID, targetID);
```
I apologize for the inconvenience here I know I was given pretty basic instructions but I seem to lack some of the understanding here of how the

**Comment by ElectScholar** [comment]
After looking further I think I have fixed most of the logic to correctly work now. I am just struggling to find the logic for the `MessageType::EMOTE_PLAYED` it is an already existing id of '371' on `game.h` but when I use it in my header the action doesn't seem to be played.

`GameMessages.h`
```cpp
	struct EmotePlayed : public GameMsg {
		EmotePlayed(int32_t emoteID, LWOOBJID targetID)
			: GameMsg(MessageType::Game::EMOTE_PLAYED), emoteID(emoteID), targetID(targetID) {}
	
		void Serialize(RakNet::BitStream& stream) const;
	
		int32_t emoteID;
		LWOOBJID targetID;
	};
```

also removing [RenderComponent::PlayAnimation](https://github.com/DarkflameUniverse/DarkflameServer/blob/main/dGame/dGameMessages/GameMessages.cpp#L4979) removes the animation part of the emote. Here is my [implementation](https://github.com/ElectScholar/DarkflameServer/blob/issue/1342/emoteMessages/dGame/dGameMessages/GameMessages.cpp#L4960-#L5007) after serializing.

**Comment by EmosewaMC** [comment]
> After looking further I think I have fixed most of the logic to correctly work now. I am just struggling to find the logic for the `MessageType::EMOTE_PLAYED` it is an already existing id of '371' on `game.h` but when I use it in my header the action doesn't seem to be played.
> 
> `GameMessages.h`
> 
> 	struct EmotePlayed : public GameMsg {
> 		EmotePlayed(int32_t emoteID, LWOOBJID targetID)
> 			: GameMsg(MessageType::Game::EMOTE_PLAYED), emoteID(emoteID), targetID(targetID) {}
> 	
> 		void Serialize(RakNet::BitStream& stream) const;
> 	
> 		int32_t emoteID;
> 		LWOOBJID targetID;
> 	};
> also removing [RenderComponent::PlayAnimation](https://github.com/DarkflameUniverse/DarkflameServer/blob/main/dGame/dGameMessages/GameMessages.cpp#L4979) removes the animation part of the emote. Here is my [implementation](https://github.com/ElectScholar/DarkflameServer/blob/issue/1342/emoteMessages/dGame/dGameMessages/GameMessages.cpp#L4960-#L5007) after serializing.

The design paradigm we generally follow is
```cpp
GameMessages::Msg msg; // notice there is no constructor called here
msg.target = targetID;
msg.var = 1;
msg.var2 = 3;
msg.Send();
```

which makes it easy to follow what is going on with the msg. This is done because these messages will be generated at some point and constructor arguments can get very lengthy vs direct assignment; as such constructors will be made with no arguments.

your issue is you arent giving the message a target via **msg.target.**  The `targetID` of the derived EmotePlayed message is who the emote is targetted at and `target` of the base GameMsg struct is the entity that is playing the emote.  In this case, we are sending the message to all connected clients saying that `target` played `emoteID` to `targetID`.

Do that and it should work.

You are using the correct msgID of EMOTE_PLAYED from the Game enum.  Don't forget to mark Serialize with override as that will fail to build on mac without it.

**Comment by ElectScholar** [comment]
@EmosewaMC Thank you for your guidance. After implementing your suggestions, the feature now works correctly for players emoting with each other. I plan to submit a pull request later today. I’m looking forward to working on similar tasks in the future, as this has really piqued my interest.

**Comment by EmosewaMC** [comment]
very good!  glad it worked for you and that you are enjoying working on the project.  :)
